### PR TITLE
Feature/eb bug fix

### LINKED ISSUE #7: [🖍] Home Interface : Title Font
[Some fonts](https://www.fontspace.com/category/Dungeons-and-Dragons)
[Some more fonts](https://www.dafont.com/theme.php?cat=405)

Screen Shot 2022-01-05 at 17 51 33

Screen Shot 2022-01-05 at 17 48 57

Screen Shot 2022-01-05 at 17 50 22

### LINKED ISSUE #8: [🖍] Home Interface : Browse Spells Link Icon
Screen Shot 2022-01-05 at 18 01 09

Screen Shot 2022-01-05 at 18 01 37

### LINKED ISSUE #9: [🖍] Home Interface : View Spellbook Link Icon
Screen Shot 2022-01-05 at 18 04 07

Screen Shot 2022-01-05 at 18 04 33

### LINKED ISSUE #18: [🐞] Bug : Displaying Character Class Icon
Screen Shot 2022-01-06 at 21 43 05

The CharClass component is being rendered by a map through all of the classes returned from a fetch call to `https://www.dnd5eapi.co/api/classes/`.  Every class has a different index that I would like to use to grab the appropriate icon from the charClassIcons object.  For example, like what is shown above, `charClassIcons.bard` finds the `bard` key is the object, which has been assigned the value of the imported icon.  The `src` attribute for the `img` tag, is expecting either a URL in the form of a string, wrapped in quotes, OR an imported icon name wrapped in curly brackets.

I have tried a few different forms of interpolation to get my desired result, but have not succeeded.

### LINKED ISSUE #19: [⚠️] Error Handling : Loading
As a user, I should see an indication that data is still loading, so that I can wait for the page to load.

dice roll

### LINKED ISSUE #46: [🧩] Feature : Local Storage
Implement local storage to keep track of spells that have been bookmarked

### PR DESCRIPTION
**Pull request by elisebeall**
## [DnD Spells]

- [x] - Feature PR [🧩]
- [x] - Refactor PR [⚙️]
- [x] - BugFix PR [🐞]
- [ ] - User Interface PR [🖍]
- [ ] - Error Handling PR [⚠️]
- [ ] - Testing PR [🧑‍💻]

**What (if any) features are you implementing?**

 - Fixed bug that was causing spells to be bookmarked multiple times.
 - Fixed bug that was keeping spells from maintaining state.
 - Bookmarked spells are now saved in local storage so they remain on page reload.
 - Added an animated loading icon that appears when the API is fetching data.
 - Closes #46 
 - Closes #7 
 - Closes #8 
 - Closes #9 
 - Closes #18 
 - Closes #19 

**What (if anything) did you refactor?**

 - Refactored UserSpellbookContext to use a reducer for adding and removing spells.
 - Refactored the Bookmark component to dispatch an action upon bookmark click 

**Were there any issues that arose?**

 - Animated gifs do not like to have transparent backgrounds.

**Is there anything needs to be prioritized/errors to be addressed?**

- No console errors.  No bugs found at this time.

**Goals? Next steps?**

 - I would like to implement the Character Selector feature so that a user is able to choose a custom character to add the bookmark.

**Current Screenshots**

https://user-images.githubusercontent.com/724355/149231635-af928c6f-9520-42d7-8b10-de30a603570c.mov

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
     "react-icons": "^4.3.1",
     "react-router-dom": "^6.2.1",
     "react-scripts": "5.0.0",
+    "uuid": "^8.3.2",
```
Comment (augustreid) [inline package.json:14]: Glad you remembered this package for creating keys!

**Review by bpeterson2579 — COMMENTED** [review]
Looks good to me

File: src/components/Bookmark.js [hunk 2]
```diff
@@ -5,29 +5,18 @@ import { FaRegBookmark as InactiveBookmark } from 'react-icons/fa';
 import { UserSpellbookContext } from '../contexts/UserSpellbookContext';
 
 const Bookmark = ({ spell }) => {
-  const [ isBookmarked, setIsBookmarked ] = useState(false);
-  const { bookmarks, setBookmarks } = useContext(UserSpellbookContext);
+  const { bookmarks, dispatch } = useContext(UserSpellbookContext);
```
Comment (augustreid) [inline src/components/Bookmark.js:8]: I was scared of reducers, but I'm glad you went for it! The result is much cleaner (and actually works lol).

File: src/components/CharSelector.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import '../css/CharSelector.css';
+import { useState, useContext } from 'react';
+import CharContext from '../contexts/CharContext';
+
+const CharSelector = () => {
+  const { characters } = useContext(CharContext);
+  const [charSelected, setCharSelected] = useState(characters[0]);
+
+  return(
```
Comment (augustreid) [inline src/components/CharSelector.js:9]: This is awesome. We should definitely get it implemented in the future!

File: src/components/ClassFilter.js [hunk 4]
```diff
@@ -1,21 +1,21 @@
...
-            <NavLink to={"/ranger/spells"} className="classLink">Ranger</NavLink>
-            <NavLink to={"/sorcerer/spells"} className="classLink">Sorcerer</NavLink>
-            <NavLink to={"/warlock/spells"} className="classLink">Warlock</NavLink>
-            <NavLink to={"/wizard/spells"} className="classLink">Wizard</NavLink>
-            <NavLink to={"/class"} className="classLink">All Classes</NavLink>
-        </nav>
-    )
+  return(
+    <nav className="classNav">
+      <NavLink to={'/bard/spells'} className="classLink">Bard</NavLink>
```
Comment (augustreid) [inline src/components/ClassFilter.js:8]: Sorry I put you through so much quotation correcting.

File: src/components/Loading.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import '../css/Loading.css';
+import loading from '../assets/squiz.gif';
+
```
Comment (augustreid) [inline src/components/Loading.js:3]: Love the idea of just making a Loading component. Good call.

File: src/components/SpellDetails.js [hunk 6]
```diff
@@ -3,25 +3,27 @@ import { useParams } from 'react-router-dom'
...
 import Bookmark from './Bookmark';
+import Loading from './Loading';
 import useFetch from '../hooks/useFetch';
+import endpoints from '../endpoints.js';
 
-const SpellDetails = ({}) => {
+const SpellDetails = () => {
   const spellIndex = useParams().index;
-  const { data: details, isLoading, error } = useFetch(`https://www.dnd5eapi.co/api/spells/${spellIndex}`)
+  const { data: details, isLoading, error } = useFetch(endpoints.spells+`${spellIndex}`)
```
Comment (augustreid) [inline src/components/SpellDetails.js:12]: Nice swap to the endpoints

File: src/components/Bookmark.js (hunk 2 above)
Comment (elisebeall) [reply src/components/Bookmark.js:8]: They are confusing, but I think I'm starting to get the hang of them!

File: src/components/CharSelector.js (hunk 3 above)
Comment (elisebeall) [reply src/components/CharSelector.js:9]: Oh yes, that file snuck it's way in there...

File: src/contexts/UserSpellbookContext.js [hunk 7]
```diff
@@ -1,15 +1,23 @@
-import { createContext, useState } from "react";
+import { createContext, useReducer, useEffect } from 'react';
+import { bookmarkReducer } from '../reducers/bookmarkReducer';
 
 export const UserSpellbookContext = createContext();
 
 const UserSpellbookContextProvider = (props) => {
-  const [ bookmarks, setBookmarks ] = useState([]);
+  const [ bookmarks, dispatch ] = useReducer(bookmarkReducer, [], () => {
```
Comment (augustreid) [inline src/contexts/UserSpellbookContext.js:7]: This is super cool. I feel like using Context together with local storage makes a lot of sense.

File: src/css/ClassFilter.css [hunk 8]
```diff
@@ -2,15 +2,15 @@
     display: flex;
     flex-direction: column;
     align-items: center;
-    background-color: #511012; 
+    background-color: #511012;
     width: 20%;
-
+    font-family: 'oblivion';
```
Comment (augustreid) [inline src/css/ClassFilter.css:7]: Thanks for catching this and adding the font to the right element.

**Review by augustreid — APPROVED** [review]
Elise, you're a bamf. Supreme monarch of squirrels and squizards alike.

**Pull request merged**